scipy/signal/_upfirdn_apply.pyx

```diff
@@ -65,6 +65,56 @@ def _output_len(Py_ssize_t len_h,
     return need
 
 
+@cython.cdivision(True)  # faster modulo
+@cython.boundscheck(False)  # designed to stay within bounds
+@cython.wraparound(False)  # we don't use negative indexing
+def _rat(x, tol=None):
```

Comment:
(Pushed a commit to fix the `while` loop limit bug -- it was because `frac_part` could get so small that `int(1. / frac_part)` (in C) would overflow and become negative). Now we get:
```
>>> from scipy.signal._upfirdn_apply import _rat
>>> _rat(10000000001/10000000000.0)
array([1, 1])
```